### PR TITLE
Status bar widget to display boolean settings state for current URL

### PR DESCRIPTION
**Pull request by toofar**
Adds a statusbar widget to show the state of boolean settings for the current URL.

Like in this poor quality picture settings-widget (`S` = javascript, `L` = geolocation)

What settings are displayed and how is configurable. This works but is ugly, I am opening this PR before it is finished in case someone has a better idea of how to do it. What I have currently done is added a new possible value for the `statusbar.widgets` settings, which is necessary in any case, and I allowed configuration of the new widget inline. For example

    `settings(S|content.javascript.enabled,G|content.webgl,L|content.geolocation,)`

Which works and isn't even breaking anything. But to get that to work I had to change the "valid values" mechanism for settings to be able to take a `type` parameter (see first commit). But due to the `settings.asciidoc` file being generated not directly from `configdata.yml` but from the configtypes that are parsed from that it ends up getting a complied regular expression's `repr` in the user documentation, which is misleading.

TL;DR: I am probably going to change from that inline configuration to just adding a new config option, so some questions:

* adding a new config option rather than doing a bunch of workarounds to allow inline configuration is the right thing to do, right?
* what shall I call it? `statusbar.settings.widget.settings`? `settings.widget.content`?
* I'll probably go with it taking a value like `'["S|content.javascript.enabled","indicator|setting-name",...]'`, anyone got any better ideas?

TODO:
- [x] make it work
- [ ] tests tests tests
- [ ] use `testAttribute()` for BoolAsk type settings where possible

### CODE REVIEW

File: qutebrowser/config/configdata.yml [hunk 1]
```diff
@@ -1291,10 +1291,22 @@ statusbar.widgets:
...
   desc: List of widgets displayed in the statusbar.
 
+settings.widget.content:
+  type:
+    name: List
+    valtype:
+      name: String
+    none_ok: true
+  default: []
+  desc: Which boolean settings to display in the statusbar settings widget.
```
Comment (mschilli87) [inline qutebrowser/config/configdata.yml]: Other multiline-description are opened by `desc: >-` and continued on the next line. I'd suggest to do the same here for consistency.

**Comment by zestyr** [comment]
Relevant issue: #1052

File: qutebrowser/mainwindow/statusbar/settings.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+from qutebrowser.mainwindow.statusbar import textbase
+from qutebrowser.utils import objreg, message
+
+
+class BooleanSettings(textbase.TextBase):
+
+    """Boolean settings state displayed in the statusbar."""
+
+    # A list of indicator, setting_name pairs.
+    _config = []
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/settings.py]: This should be set in `__init__` instead, or all instances of the class will share it, which might be quite confusing.

File: qutebrowser/mainwindow/statusbar/settings.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        for part in config_:
+            try:
+                _indicater, value = part
+                self._to_bool(value, None)
+            except ValueError as err:
+                message.error(
+                    "Error parsing '{}' setting value {}: {}"
+                    .format(self.config_option, part, err)
+                )
+                return
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/settings.py]: Like above, this kind of validation should probably be the config's responsibility.

File: qutebrowser/mainwindow/statusbar/settings.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        self._config = config_
+        tab = self._current_tab()
+        if tab:
+            self.on_url_changed(tab.url())
+
+    def on_url_changed(self, url):
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/settings.py]: All those `on_*` methods should probably have `@pyqtSlot(...)`

File: qutebrowser/mainwindow/statusbar/settings.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    def on_load_status_changed(self, _status):
+        """Update widget when load status changed."""
+        self.on_tab_changed(self._current_tab())
+
+    def _current_tab(self):
+        """Get the currently displayed tab."""
+        # pylint: disable=protected-access
+        window = objreg.get('tabbed-browser', scope='window',
+                            window=self.parent()._win_id)
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/settings.py]: It's probably cleaner to pass the `win_id` to `BooleanSettings` when creating it instead.

File: qutebrowser/mainwindow/statusbar/settings.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        )
+
+    def parse_config(self):
+        """Parse settings from the config option.
+
+        A ValueError will be raised for any errors in parsing or
+        applying the settings.
+        """
+        raw = config.instance.get(self.config_option)
+        config_ = [x.split('|') for x in raw if x.strip()]
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/settings.py]: might want to create `collections.namedtuple` instances here so you can do `part.indicator` and `part.value` elsewhere.

File: qutebrowser/mainwindow/statusbar/settings.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        # Use `_to_bool()` to do some early validation
+        for part in config_:
+            try:
+                _indicater, value = part
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/settings.py]: _indicater -> _indicator

File: qutebrowser/browser/browsertab.py [hunk 8]
```diff
@@ -1016,3 +1020,25 @@ def __repr__(self):
...
+        """Return true if the user has granted permission for `setting_name`.
+
+        Returns KeyError if `setting_name` doesn't map to a grantable
+        feature.
+        """
+        try:
+            feat = [
+                f for f in self._permissions.features.values()
+                if f.setting_name == setting_name
+            ][0]
```
[the remark this answers is not in the document]
Comment (toofar) [reply qutebrowser/browser/browsertab.py]: Swittching this to build a dict revealed another problem. The `content.media_capture` setting maps to three features, `MediaAudioCapture`, `MediaVideoCapture` and `MediaAudioVideoCapture`. (Same for `content.desktop_capture`.) When I was using a list comprehension it was returning the state of the first one (audio capture) which just so happened to be the one I was testing with.
I have kept it as a list comprehension for now and am using `any()`.

File: qutebrowser/mainwindow/statusbar/settings.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            self.on_url_changed(tab.url())
+
+    @pyqtSlot(QUrl)
+    def on_url_changed(self, url):
+        """Update the widget to reflect settings for url."""
+        if not url.isValid():
+            url = None
+        parts = [
+            indicator for setting, indicator in self._config.items()
+            if self._to_bool(setting, url)
```
[the remark this answers is not in the document]
Comment (toofar) [reply qutebrowser/mainwindow/statusbar/settings.py]: On 5.11 and 5.7 it gets called on `LoadStatus.loading` and `LoadStatus.success_https`, on 5.12 the second one comes twice. We could potentially only bother updating on the `LoadStatus.loading` status but I don't know for sure if we will always get that (it does fire for everything I thought to try including `data:` URLs though). Caching it sure easier to reason about.
